src/java/org/apache/cassandra/config/DataRateSpec.java

```diff
@@ -27,99 +27,88 @@
     }
 
-    DataRateSpec(double quantity, DataRateUnit unit)
+    private DataRateSpec(String value, DataRateUnit minUnit, long max)
     {
-        if (quantity < 0)
-            throw new ConfigurationException("Invalid bit rate: value must be non-negative");
+        this(value);
 
-        if (quantity > Long.MAX_VALUE)
-            throw new NumberFormatException("Invalid bit rate: value must be between 0 and Long.MAX_VALUE = 9223372036854775807");
+        validateQuantity(value, quantity(), unit(), minUnit, max);
 
-        this.quantity = quantity;
-        this.unit = unit;
     }
 
-    /**
-     * Creates a {@code DataRateSpec} of the specified amount of bits per second.
-     *
-     * @param bytesPerSecond the amount of bytes per second
-     * @return a {@code DataRateSpec}
-     */
-    public static DataRateSpec inBytesPerSecond(long bytesPerSecond)
+    private DataRateSpec(double quantity, DataRateUnit unit, DataRateUnit minUnit, long max)
     {
-        return new DataRateSpec(bytesPerSecond, DataRateUnit.BYTES_PER_SECOND);
+        this.quantity = quantity;
+        this.unit = unit;
+
+        validateQuantity(quantity, unit, minUnit, max);
     }
 
-    /**
-     * Creates a {@code DataRateSpec} of the specified amount of kibibytes per second.
-     *
-     * @param kibibytesPerSecond the amount of kibibytes per second
-     * @return a {@code DataRateSpec}
-     */
-    public static DataRateSpec inKibibytesPerSecond(long kibibytesPerSecond)
+    private static void validateQuantity(String value, double quantity, DataRateUnit unit, DataRateUnit minUnit, long max)
     {
-        return new DataRateSpec(kibibytesPerSecond, DataRateUnit.KIBIBYTES_PER_SECOND);
+        // negatives are not allowed by the regex pattern
+        if (minUnit.convert(quantity, unit) >= max)
+            throw new ConfigurationException("Invalid data rate: " + value + ". It shouldn't be more than " +
+                                             (max - 1) + " in " + minUnit.name().toLowerCase());
     }
 
-    /**
-     * Creates a {@code DataRateSpec} of the specified amount of mebibytes per second.
-     *
-     * @param mebibytesPerSecond the amount of mebibytes per second
-     * @return a {@code DataRateSpec}
-     */
-    public static DataRateSpec inMebibytesPerSecond(long mebibytesPerSecond)
+    private static void validateQuantity(double quantity, DataRateUnit unit, DataRateUnit minUnit, long max)
     {
-        return new DataRateSpec(mebibytesPerSecond, DataRateUnit.MEBIBYTES_PER_SECOND);
+        if (quantity < 0)
+            throw new ConfigurationException("Invalid data rate: value must be non-negative");
+
+        if (minUnit.convert(quantity, unit) >= max)
+            throw new ConfigurationException("Invalid data rate: " + quantity + " " + unit.name().toLowerCase() + ". It shouldn't be more than " +
+                                             (max - 1) + " in " + minUnit.name().toLowerCase());
```

Comment:
Nit: we could use `String.format`:
```suggestion
           throw new ConfigurationException(String.format("Invalid data rate: %s %s. It shouldn't be more than %d in %s",
                                                           quantity, unit.name().toLowerCase(),
                                                           max - 1, minUnit.name().toLowerCase()));
```